src/java/org/apache/cassandra/security/DefaultSslContextFactoryImpl.java

```diff
@@ -0,0 +1,248 @@
+        try
+        {
+            sslContext = builder.build();
+        }
+        catch (SSLException e)
+        {
+            throw new SSLException("failed to build the final SslContext object for secure connections", e);
+        }
+        return sslContext;
+    }
+
+    @Override
+    public synchronized void initHotReloading(EncryptionOptions options) throws SSLException {
+
+        List<HotReloadableFile> fileList = new ArrayList<>();
+
+        if (options != null && options.tlsEncryptionPolicy() != EncryptionOptions.TlsEncryptionPolicy.UNENCRYPTED)
+        {
+            fileList.add(new HotReloadableFile(options.keystore));
+            fileList.add(new HotReloadableFile(options.truststore));
+        }
+
+        hotReloadableFiles = ImmutableList.copyOf(fileList);
+    }
+
+    @Override
+    public boolean shouldReload()
+    {
+        return hotReloadableFiles.stream().anyMatch(HotReloadableFile::shouldReload);
+    }
+
+    @Override
+    public boolean hasKeystore(EncryptionOptions options) {
+        return new File(options.keystore).exists();
+    }
+
+    /**
+     * Helper class for hot reloading SSL Contexts
+     */
+    private static class HotReloadableFile
+    {
+        private final File file;
+        private volatile long lastModTime;
+
+        HotReloadableFile(String path)
+        {
+            file = new File(path);
+            lastModTime = file.lastModified();
+        }
+
+        boolean shouldReload()
+        {
+            long curModTime = file.lastModified();
+            boolean result = curModTime != lastModTime;
+            lastModTime = curModTime;
+            return result;
+        }
+
+        @Override
+        public String toString()
+        {
+            return "HotReloadableFile{" +
+                   "file=" + file +
+                   ", lastModTime=" + lastModTime +
+                   '}';
+        }
+    }
+
+    KeyManagerFactory buildKeyManagerFactory(EncryptionOptions options) throws SSLException
+    {
+        try (InputStream ksf = Files.newInputStream(Paths.get(options.keystore)))
+        {
+            KeyManagerFactory kmf = KeyManagerFactory.getInstance(
+            options.algorithm == null ? KeyManagerFactory.getDefaultAlgorithm() : options.algorithm);
+            KeyStore ks = KeyStore.getInstance(options.store_type);
+            ks.load(ksf, options.keystore_password.toCharArray());
+            if (!checkedExpiry)
+            {
+                for (Enumeration<String> aliases = ks.aliases(); aliases.hasMoreElements(); )
+                {
+                    String alias = aliases.nextElement();
+                    if (ks.getCertificate(alias).getType().equals("X.509"))
+                    {
+                        Date expires = ((X509Certificate) ks.getCertificate(alias)).getNotAfter();
+                        if (expires.before(new Date()))
+                            logger.warn("Certificate for {} expired on {}", alias, expires);
```

Comment:
we just warn and that is it? are we ok with using expired certificate?

Reply:
+1 it deserves some thought. Being expired it will at least be an encrypted connection. Idk how desirable it would be for a cluster to suddenly stop working. In many places I worked before people onlu notice a cert expired when a system stopped working. I don't know if we want to be that strict :thinking: Strictly speaking yes, but more opinions welcomed...

Reply:
Ok so this comes from [here](https://github.com/apache/cassandra/pull/1027/files#diff-0d569398cfd58566fc56bfb80c971a72afe3f392addc2df731a0b44baf29019eL242) iiuc. Hence this is the current and accepted behavior which settles it for me.